src/Nethermind/Nethermind.Optimism/OptimismEIP1559Parameters.cs

```diff
@@ -25,69 +28,128 @@ public EIP1559Parameters(byte version, UInt32 denominator, UInt32 elasticity)
+        if (version > maxVersion)
+        {
+            error = $"{nameof(version)} must be between 0 and {maxVersion}";
+            return false;
+        }
+
+        var expLength = EIP1559Parameters.ByteLengthByVersion[version];
+        if (dataLength != expLength)
         {
-            parameters = default;
-            error = $"{nameof(header.ExtraData)} data must be {EIP1559Parameters.ByteLength} bytes long";
+            error = $"{nameof(header.ExtraData)} must be {expLength} bytes long";
             return false;
         }
 
-        ReadOnlySpan<byte> extraData = header.ExtraData.AsSpan();
-        var version = extraData.TakeAndMove(1)[0];
-        var denominator = BinaryPrimitives.ReadUInt32BigEndian(extraData.TakeAndMove(4));
-        var elasticity = BinaryPrimitives.ReadUInt32BigEndian(extraData.TakeAndMove(4));
+        var denominator = BinaryPrimitives.ReadUInt32BigEndian(data.TakeAndMove(sizeof(UInt32)));
+        var elasticity = BinaryPrimitives.ReadUInt32BigEndian(data.TakeAndMove(sizeof(UInt32)));
+
+        if (version == 0)
+            return EIP1559Parameters.TryCreateV0(denominator, elasticity, out parameters, out error);
+
+        var minBaseFee = BinaryPrimitives.ReadUInt64BigEndian(data.TakeAndMove(sizeof(UInt64)));
+        return EIP1559Parameters.TryCreateV1(denominator, elasticity, minBaseFee, out parameters, out error);
 
-        return EIP1559Parameters.TryCreate(version, denominator, elasticity, out parameters, out error);
     }
 
     public static bool TryDecodeEIP1559Parameters(this OptimismPayloadAttributes attributes, out EIP1559Parameters parameters, [NotNullWhen(false)] out string? error)
     {
-        if (attributes.EIP1559Params?.Length != 8)
+        parameters = default;
+
+        ReadOnlySpan<byte> data = attributes.EIP1559Params;
+        var dataLength = data.Length;
+        if (dataLength == 0)
         {
-            parameters = default;
-            error = $"{nameof(attributes.EIP1559Params)} must be 8 bytes long";
+            error = $"{nameof(attributes.EIP1559Params)} must not be empty";
             return false;
         }
 
-        ReadOnlySpan<byte> span = attributes.EIP1559Params.AsSpan();
-        var denominator = BinaryPrimitives.ReadUInt32BigEndian(span.TakeAndMove(4));
-        var elasticity = BinaryPrimitives.ReadUInt32BigEndian(span.TakeAndMove(4));
+        var version = Array.IndexOf(EIP1559Parameters.ByteLengthByVersion, (byte)(dataLength + 1));
+        if (version < 0)
+        {
+            error = $"{nameof(attributes.EIP1559Params)} has invalid length";
+            return false;
+        }
+
+        var denominator = BinaryPrimitives.ReadUInt32BigEndian(data.TakeAndMove(sizeof(UInt32)));
+        var elasticity = BinaryPrimitives.ReadUInt32BigEndian(data.TakeAndMove(sizeof(UInt32)));
+
+        if (version == 0)
+            return EIP1559Parameters.TryCreateV0(denominator, elasticity, out parameters, out error);
```

Comment:
It's better to surround the statement with  {} and use specific types instead of vars

Reply:
For example `var elasticity` and actually everywhere in the pr. It's much easier to read when type is mentioned. @rubo is it possible to make IDEs suggest/enforce using explicit types?